src/System.Linq/src/System/Linq/Union.cs

```diff
@@ -11,25 +12,130 @@ public static partial class Enumerable
     {
         public static IEnumerable<TSource> Union<TSource>(this IEnumerable<TSource> first, IEnumerable<TSource> second)
         {
-            if (first == null) throw Error.ArgumentNull("first");
-            if (second == null) throw Error.ArgumentNull("second");
-            return UnionIterator(first, second, null);
+            return Union(first, second, null);
         }
 
         public static IEnumerable<TSource> Union<TSource>(this IEnumerable<TSource> first, IEnumerable<TSource> second, IEqualityComparer<TSource> comparer)
         {
             if (first == null) throw Error.ArgumentNull("first");
             if (second == null) throw Error.ArgumentNull("second");
-            return UnionIterator(first, second, comparer);
+            return new UnionIterator<TSource>(first, second, comparer);
         }
 
-        private static IEnumerable<TSource> UnionIterator<TSource>(IEnumerable<TSource> first, IEnumerable<TSource> second, IEqualityComparer<TSource> comparer)
+        private sealed class UnionIterator<TSource> : Iterator<TSource>, IIListProvider<TSource>
         {
-            Set<TSource> set = new Set<TSource>(comparer);
-            foreach (TSource element in first)
-                if (set.Add(element)) yield return element;
-            foreach (TSource element in second)
-                if (set.Add(element)) yield return element;
+            private readonly IEnumerable<TSource> _first;
+            private readonly IEnumerable<TSource> _second;
+            private readonly IEqualityComparer<TSource> _comparer;
+            private Set<TSource> _set;
+            private IEnumerator<TSource> _enumerator;
+
+            public UnionIterator(IEnumerable<TSource> first, IEnumerable<TSource> second, IEqualityComparer<TSource> comparer)
+            {
+                Debug.Assert(first != null);
+                Debug.Assert(second != null);
+                _first = first;
+                _second = second;
+                _comparer = comparer;
+            }
+
+            public override Iterator<TSource> Clone()
+            {
+                return new UnionIterator<TSource>(_first, _second, _comparer);
+            }
+
+            private bool GetNext()
+            {
+                while (_enumerator.MoveNext())
+                {
+                    TSource element = _enumerator.Current;
+                    if (_set.Add(element))
+                    {
+                        current = element;
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+
+            public override bool MoveNext()
+            {
+                switch(state)
+                {
+                    case 1:
+                        _enumerator = _first.GetEnumerator();
+                        if (_enumerator.MoveNext())
+                            state = 2;
```

Comment:
I also like the corefx style more. Except for the avoidance of "var". :-) 
We should run the formatter tool at some point on Linq and Expressions just for consistency.

Entered https://github.com/dotnet/corefx/issues/6150 for this.